### ISSUE TITLE
Should extracted features be normalized before querying the FAISS index?

### ISSUE THREAD

**Issue by paranoid2droid** [utter-project/mHuBERT-147-scripts#3]
Hi,

Thank you for this excellent project and the well-documented codebase! I am working on a custom pipeline where I extract features using mHuBERT and query the FAISS index for clustering. However, I am unsure whether feature normalization is required in my usage.

Below is a snippet of how I am currently using mHuBERT to extract features and query the FAISS index:

```
with torch.no_grad():
    # Extract features using mHuBERT
    temp = mhubert.extract_features(source=wav, mask=False, output_layer=6)[0]  # Shape: (batch_size, seq_len, hidden_dim)

    # Calculate frame lengths dynamically
    lengths = [temp[i].shape[0] for i in range(temp.shape[0])]

    # Reshape features for batch FAISS search
    features = temp.reshape(-1, temp.shape[-1]).cpu().numpy()  # Flatten features

    # Normalize features (optional, unclear if needed)
    features /= np.linalg.norm(features, axis=1, keepdims=True)  # L2 normalization

    # Query the FAISS index
    _, C = get_centroids_index(features, self.faiss_index, self.faiss_ivf)

    # Map centroids to audio segments
    vecs = apply_centroids_to_audios(lengths, C)
```

However, I am not clear about the normalization of FAISS index, seeing some instructions saying that the FAISS index expects normalized features during training and usage.

Were the features normalized when training the FAISS index? Should the extracted features be normalized (e.g., L2 normalization) before querying the FAISS index to ensure consistency?

Thanks again, and I look forward to hearing your opinions!

**Comment by mzboito** [comment]
Hello,

Many thanks for the interest in the model. It is great to hear that people are using it! :) 

You do not need to normalize your features before passing it to your trained faiss index. 

The only normalization employed on the training pipeline was input normalization (https://github.com/utter-project/fairseq/blob/2959684ecc6e28e34a075e2f1a08bba100e11743/examples/mHuBERT-147/config/pretrain/mhubert147_base.yaml#L35).

**Comment by paranoid2droid** [comment]
@mzboito Thanks for the prompt reply! That's a lot of help.
